packages/product-json-to-csv/test/main.spec.js

```diff
@@ -0,0 +1,625 @@
+            "key": "mv-key"
+          },
+          "variants": []
+        }`
+      inputStream = streamTest.fromChunks([product])
+    })
+
+    it('should take an inputStream and output highland stream', () => {
+      const productStream = productJsonToCsv.parse(inputStream)
+      expect(productStream.source.__HighlandStream__).toBeTruthy()
+    })
+
+    it('should return a flattened product object', done => {
+      const productStream = productJsonToCsv.parse(inputStream)
+      const expected = [
+        {
+          id: 'product-1-id',
+          'slug.en': 'my-slug-1',
+          'variant.id': 'mv-id',
+          'variant.key': 'mv-key',
+        },
+      ]
+      productStream.toArray(actual => {
+        expect(actual).toEqual(expect.arrayContaining(expected))
+        done()
+      })
+    })
+
+    it('should log and emit error if error occurs', done => {
+      const fakeError = new Error('fake error')
+      productJsonToCsv._resolveReferences = jest.fn(() =>
+        Promise.reject(fakeError)
+      )
+      const outputStream = {
+        emit: jest.fn(),
+      }
+      const productStream = productJsonToCsv.parse(inputStream, outputStream)
+
+      // Stream is lazy and has to be consumed
+      productStream.done(() => {
+        expect(productJsonToCsv.logger.error).toBeCalledWith(fakeError)
+        expect(outputStream.emit).toBeCalledWith('error', fakeError)
+        done()
+      })
+    })
+  })
+
+  describe('Resolvers', () => {
+    let sampleProduct
+    beforeEach(() => {
+      sampleProduct = {
+        id: 'myProduct-1',
+        productType: { id: 'product-type-id' },
+        taxCategory: { id: 'tax-cat-id' },
+        state: { id: 'state-id' },
+        categories: [{ id: 'cat-id-1' }, { id: 'cat-id-2' }],
+        categoryOrderHints: { 'cat-id-1': '0.012', 'cat-id-2': '0.987' },
+      }
+    })
+
+    describe('::_resolveReferences', () => {
+      beforeEach(() => {
+        const productType = {
+          id: 'product-type-id',
+          name: 'resolved-product-type',
+          attributes: [{}],
+        }
+        const taxCategory = {
+          id: 'tax-cat-id',
+          key: 'resolved-tax-cat-key',
+        }
+        const state = {
+          id: 'state-id',
+          key: 'res-state-key',
+        }
+        const categories = [
+          { id: 'cat-id-1', name: { en: 'res-cat-name-1' } },
+          { id: 'cat-id-2', name: { en: 'res-cat-name-2' } },
+        ]
+        const categoryOrderHints = {
+          'res-cat-name-1': '0.015',
+          'res-cat-name-2': '0.987',
+        }
+
+        productJsonToCsv._resolveProductType = jest.fn(() => ({ productType }))
+        productJsonToCsv._resolveTaxCategory = jest.fn(() => ({ taxCategory }))
+        productJsonToCsv._resolveState = jest.fn(() => ({ state }))
+        productJsonToCsv._resolveCategories = jest.fn(() => ({ categories }))
+        productJsonToCsv._resolveCategoryOrderHints = jest.fn(() => ({
```

Comment:
Oh this reaching in feels fragile. To me its a bit of an indication of the module/class not following the single responsibilty pattern. If we extract this resolving things into modules themselves we can mock them without doing the reaching in.